### PR TITLE
CE-2963 Allow users to classify templates using keyboard only

### PR DESCRIPTION
**Pull request by adamkarminski**
/cc: @Wikia/community-engineering @kvas-damian

### CODE REVIEW

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 1]
```diff
@@ -75,14 +77,16 @@ function ($, mw, loader, nirvana, tracker, labeling) {
...
+		// Mark selected type
+		$preselectedType = $classificationForm.find('input#template-classification-' + templateType);
 
-			if (!!$preselectedType) {
-				$classificationForm.find('input[checked="checked"]').removeAttr('checked');
-				$preselectedType.attr('checked', 'checked');
-			}
+		if (!!$preselectedType) {
+			$classificationForm.find('input[checked="checked"]').removeAttr('checked');
+			$preselectedType.attr('checked', 'checked').prop('autofocus', true);
```
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: autofocus does nothing here due to focus set after show

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 2]
```diff
@@ -243,6 +254,24 @@ function ($, mw, loader, nirvana, tracker, labeling) {
 		return false;
 	}
 
+	function openModalKeyboardShortcut(e) {
+		if (e.shiftKey && (e.ctrlKey || e.metaKey)) {
+			var keyCode = e.keyCode ? e.keyCode : e.which;
+			if (keyCode === 75) {
```
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: key combination in comment will be useful

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 3]
```diff
@@ -243,6 +254,24 @@ function ($, mw, loader, nirvana, tracker, labeling) {
...
+				e.preventDefault();
+				openEditModal('editType');
+			}
+		}
+	}
+
+	function submitFormOnEnterKeyPress(e) {
+		var keyCode = e.keyCode ? e.keyCode : e.which;
+		if (keyCode === 13) {
```
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js:287]: explanation of 13 might be useful

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 4]
```diff
@@ -75,14 +77,16 @@ function ($, mw, loader, nirvana, tracker, labeling) {
...
 
-		if (templateType) {
-			// Mark selected type
-			$preselectedType = $classificationForm.find('input[value="' + templateType + '"]');
+		if (!templateType) {
+			templateType = 'unknown';
+		}
+
+		// Mark selected type
+		$preselectedType = $classificationForm.find('input#template-classification-' + templateType);
```
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: 'input' is not needed as id is already unique selector

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 5]
```diff
@@ -33,6 +33,8 @@ function ($, mw, loader, nirvana, tracker, labeling) {
 		typeGetter = typeGetterProvided;
 		$typeLabel = $('.template-classification-type-text');
 
+		$(w).bind('keydown', openModalKeyboardShortcut);
```
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: What about unbinding for a time that modal is open or at least double opening prevent.

Comment (Grunny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: `.on` is preferred in jQuery 1.7+.

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 6]
```diff
@@ -75,14 +77,16 @@ function ($, mw, loader, nirvana, tracker, labeling) {
 			$classificationForm = $(classificationForm[0]);
 		}
 
-		if (templateType) {
-			// Mark selected type
-			$preselectedType = $classificationForm.find('input[value="' + templateType + '"]');
+		if (!templateType) {
```
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: This brakes requirement for forced modal not to have any type selected on new template creation

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 7]
```diff
@@ -75,14 +77,16 @@ function ($, mw, loader, nirvana, tracker, labeling) {
...
+		// Mark selected type
+		$preselectedType = $classificationForm.find('#template-classification-' + templateType);
 
-			if (!!$preselectedType) {
-				$classificationForm.find('input[checked="checked"]').removeAttr('checked');
-				$preselectedType.attr('checked', 'checked');
-			}
+		if (!!$preselectedType) {
+			$classificationForm.find('input[checked="checked"]').removeAttr('checked');
+			$preselectedType.attr('checked', 'checked');
```
Comment (Grunny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js:89]: Properties such as checked should be retrieved and set using `.prop`.

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 8]
```diff
@@ -118,14 +122,16 @@ function ($, mw, loader, nirvana, tracker, labeling) {
 		modalInstance.bind('done', function runSave(e) {
 			processSave(modalInstance);
 
-			// Track - primary-button click
+			// Track - primary-button click or save by pressing Enter key
+			var label = e ? $(e.currentTarget).text() : 'keypress';
```
Comment (Grunny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: Variable declarations should be at the top of the scope context [per conventions](https://github.com/Wikia/guidelines/blob/master/JavaScript/CodingConventions.md#variables).

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 9]
```diff
@@ -75,14 +77,16 @@ function ($, mw, loader, nirvana, tracker, labeling) {
...
+		}
+
+		// Mark selected type
+		$preselectedType = $classificationForm.find('#template-classification-' + templateType);
 
-			if (!!$preselectedType) {
-				$classificationForm.find('input[checked="checked"]').removeAttr('checked');
-				$preselectedType.attr('checked', 'checked');
-			}
+		if (!!$preselectedType) {
```
Comment (Grunny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: `if ($preselectedType.length) {` would be a more standard check.

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 10]
```diff
@@ -243,6 +254,26 @@ function ($, mw, loader, nirvana, tracker, labeling) {
 		return false;
 	}
 
+	function openModalKeyboardShortcut(e) {
+		// Shortcut - Action Key (Ctrl or Cmd) + Shift + K
+		if (e.shiftKey && (e.ctrlKey || e.metaKey)) {
+			var keyCode = e.keyCode ? e.keyCode : e.which;
```
Comment (Grunny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: Same here for the variable declaration.

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 11]
```diff
@@ -141,8 +147,13 @@ function ($, mw, loader, nirvana, tracker, labeling) {
 			});
 		});
 
+		$(w).bind('keypress', {modalInstance: modalInstance}, submitFormOnEnterKeyPress);
```
Comment (kvas-damian) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: $(w) is used many times in this module. Maybe we can store it in a variable?

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 12]
```diff
@@ -141,8 +147,13 @@ function ($, mw, loader, nirvana, tracker, labeling) {
...
+
 		/* Show the modal */
 		modalInstance.show();
+
+		// Make sure that focus is in the right place
+		$('#template-classification-' + mw.html.escape($preselectedType.val())).focus();
 	}
 
 	function processSave(modalInstance) {
```
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js:164]: Can we distinguish tracking submit via keyboard from regular click 
[TemplateClassificationModal.js#L170](https://github.com/Wikia/app/blob/CE-2963/extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js#L170)
like we do on closing modal?

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 13]
```diff
@@ -243,6 +254,26 @@ function ($, mw, loader, nirvana, tracker, labeling) {
 		return false;
 	}
 
+	function openModalKeyboardShortcut(e) {
+		// Shortcut - Action Key (Ctrl or Cmd) + Shift + K
+		if (e.shiftKey && (e.ctrlKey || e.metaKey)) {
+			var keyCode = e.keyCode ? e.keyCode : e.which;
+			if (keyCode === 75) {
+				e.preventDefault();
+				openEditModal('editType');
```
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: What about tracking opening modal with keyboard shortcut?

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js (hunk 12 above)
Comment (adamkarminski) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js:164]: We **do** track submit (or `done` actually), not closing. https://github.com/Wikia/app/blob/CE-2963/extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js#L122-L131

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js (hunk 11 above)
Comment (adamkarminski) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: Maybe we can! :)

**Comment by kamilkoterba** [comment]
These are not clicks only anymore
[TemplateClassificationModal.js#L170](https://github.com/Wikia/app/blob/CE-2963/extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js#L170)
[TemplateClassificationModal.js#L145](https://github.com/Wikia/app/blob/CE-2963/extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js#L145)

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js (hunk 7 above)
Comment (adamkarminski) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js:89]: Agreed. :)

**Comment by adamkarminski** [comment]
@kamilkoterba - I am not sure if we actually care. I'll ask Trevor if he actually wants to track keyboard-specific actions.

**Comment by adamkarminski** [comment]
@kamilkoterba - Actually, tracking keyboard actions is not part of the acceptance criteria. Can we assume that we don't change anything for now and will do the tracking as a separate story if there's a need?

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js (hunk 12 above)
Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js:164]: :+1:

**Comment by kamilkoterba** [comment]
That's good approach. We can open a ticket for improvements if such criteria appears.

LGTM

**Pull request merged**